File: ExerciseCategory.cpp
#include "ExerciseCategory.hpp"
#include <algorithm>
#include <sstream>

CategoryBase::CategoryBase(const std::string& name) : categoryName(name) {}

CategoryBase::CategoryBase(const CategoryBase& other) : categoryName(other.categoryName) {}

CategoryBase::~CategoryBase() = default;

std::string CategoryBase::getCategoryName() const {
    return categoryName;
}

ExerciseCategory::ExerciseCategory() : CategoryBase("Default Category"), description(std::make_unique<std::string>("No description")) {}

ExerciseCategory::ExerciseCategory(const std::string& name, const std::string& desc)
    : CategoryBase(name), description(std::make_unique<std::string>(desc)) {}

ExerciseCategory::ExerciseCategory(const ExerciseCategory& other)
    : CategoryBase(other), exercises(other.exercises), 
      description(other.description ? std::make_unique<std::string>(*other.description) : nullptr) {}

ExerciseCategory& ExerciseCategory::operator=(const ExerciseCategory& other) {
    if (this != &other) {
        CategoryBase::operator=(other);
        exercises = other.exercises;
        if (other.description) {
            description = std::make_unique<std::string>(*other.description);
        } else {
            description.reset();
        }
    }
    return *this;
}

ExerciseCategory::ExerciseCategory(ExerciseCategory&& other) noexcept
    : CategoryBase(std::move(other)), exercises(std::move(other.exercises)), 
      description(std::move(other.description)) {}

ExerciseCategory& ExerciseCategory::operator=(ExerciseCategory&& other) noexcept {
    if (this != &other) {
        CategoryBase::operator=(std::move(other));
        exercises = std::move(other.exercises);
        description = std::move(other.description);
    }
    return *this;
}

ExerciseCategory::~ExerciseCategory() = default;

std::vector<Exercise> ExerciseCategory::filterByGoal(const std::string& goal, const std::set<std::string>& equipment, const std::string& level) const {
    std::vector<Exercise> result;
    for (const auto& ex : exercises) {
        if (ex->getType().find(goal) != std::string::npos) {
            result.push_back(*ex);
        }
    }
    return result;
}

std::vector<Exercise> ExerciseCategory::recommendForSlot(const std::string& workoutSlot, const UserProfile& user, const std::string& history) const {
    std::vector<Exercise> result;
    std::string searchTerm = workoutSlot + " " + user.getProfileInfo();
    size_t pos = searchTerm.find(history);
    if (pos != std::string::npos) {
        searchTerm.replace(pos, history.length(), "recommended");
    }
    std::stringstream ss;
    ss << "Recommendations for: " << searchTerm;
    std::string message = ss.str();
    (void)message;

    for (const auto& ex : exercises) {
        if (ex->getType() == "Exercise") {
            result.push_back(*ex);
        }
    }
    return result;
}

bool ExerciseCategory::operator==(const ExerciseCategory& other) const {
    return getCategoryName() == other.getCategoryName();
}

bool ExerciseCategory::operator!=(const ExerciseCategory& other) const {
    return !(*this == other);
}

ExerciseCategory ExerciseCategory::operator+(const ExerciseCategory& other) const {
    ExerciseCategory combined(getCategoryName() + " & " + other.getCategoryName(), *description + " + " + *other.description);
    combined.exercises.reserve(exercises.size() + other.exercises.size());
    combined.exercises.insert(combined.exercises.end(), exercises.begin(), exercises.end());
    combined.exercises.insert(combined.exercises.end(), other.exercises.begin(), other.exercises.end());
    return combined;
}

std::ostream& operator<<(std::ostream& os, const ExerciseCategory& cat) {
    os << "Category: " << cat.getCategoryName();
    if (cat.description) {
        os << " - " << *cat.description;
    }
    os << "\nExercises: " << cat.exercises.size() << "\n";
    return os;
}

File: ExerciseCategory.java
import java.util.ArrayList;
import java.util.List;

class CategoryBase {
    protected String categoryName;

    public CategoryBase() {
        this("General");
    }

    public CategoryBase(String name) {
        this.categoryName = name;
    }

    public CategoryBase(CategoryBase other) {
        this.categoryName = other.categoryName;
    }

    public String getCategoryName() {
        return categoryName;
    }
}

public class ExerciseCategory extends CategoryBase {
    private List<Exercise> exercises = new ArrayList<>();
    private String description;

    public ExerciseCategory() {
        super("Default Category");
        this.description = "No description";
    }

    public ExerciseCategory(String name, String desc) {
        super(name);
        this.description = desc;
    }

    public ExerciseCategory(ExerciseCategory other) {
        super(other);
        this.exercises = new ArrayList<>(other.exercises);
        this.description = (other.description != null) ? new String(other.description) : null;
    }

    public ExerciseCategory assign(ExerciseCategory other) {
        if (this != other) {
            super.categoryName = other.categoryName;
            this.exercises = new ArrayList<>(other.exercises);
            this.description = (other.description != null) ? new String(other.description) : null;
        }
        return this;
    }

    public List<Exercise> filterByGoal(String goal, Set<String> equipment, String level) {
        List<Exercise> result = new ArrayList<>();
        for (Exercise ex : exercises) {
            if (ex.getType().contains(goal)) {
                result.add(new Exercise(ex));
            }
        }
        return result;
    }

    public List<Exercise> recommendForSlot(String workoutSlot, UserProfile user, String history) {
        List<Exercise> result = new ArrayList<>();
        String searchTerm = workoutSlot + " " + user.getProfileInfo();
        int pos = searchTerm.indexOf(history);
        if (pos != -1) {
            searchTerm = searchTerm.substring(0, pos) + "recommended" + searchTerm.substring(pos + history.length());
        }
        StringBuilder sb = new StringBuilder();
        sb.append("Recommendations for: ").append(searchTerm);
        String message = sb.toString();
        for (Exercise ex : exercises) {
            if (ex.getType().equals("Exercise")) {
                result.add(new Exercise(ex));
            }
        }
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        ExerciseCategory other = (ExerciseCategory) obj;
        return getCategoryName().equals(other.getCategoryName());
    }

    public boolean notEquals(ExerciseCategory other) {
        return !this.equals(other);
    }

    public ExerciseCategory add(ExerciseCategory other) {
        ExerciseCategory combined = new ExerciseCategory(getCategoryName() + " & " + other.getCategoryName(), description + " + " + other.description);
        combined.exercises.addAll(exercises);
        combined.exercises.addAll(other.exercises);
        return combined;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("Category: ").append(getCategoryName());
        if (description != null) {
            sb.append(" - ").append(description);
        }
        sb.append("\nExercises: ").append(exercises.size()).append("\n");
        return sb.toString();
    }
}